task 5

vershininm opened the pull request.

vershininm committed:
task-5: re-generate Flusher mock

vershininm committed:
task-5: introduce and implement Saver interface (including its tests)

applejohnny commented on internal/saver/saver.go:
```diff
@@ -0,0 +1,131 @@
+)
+
+const (
+	inited = iota + 1
```
Тут кроется потенциальная ошибка: добавление ещё одного стейта приведёт к ошибке, т.к. `iota = 3`  это первые два флага.
Константы-флаги лучше обозначать через бинарные литерылы `0b01`. Так ещё сразу понятно, что это набор флагов, а не перечисление.

applejohnny commented on internal/saver/saver.go:
```diff
@@ -0,0 +1,131 @@
+			case req, ok := <-s.flushQueue:
+				if !ok {
+					requests = s.flush(requests)
+					s.wait.Done()
```
`wait.Done()` обычно ставится через `defer` вначале функции

applejohnny commented on internal/saver/saver.go:
```diff
@@ -0,0 +1,131 @@
+			select {
+			case req, ok := <-s.flushQueue:
+				if !ok {
+					requests = s.flush(requests)
```
Неочевидный способ очистки `requests`.
Более гошный стиль это:
`s.flush(requests)`
`requests = requests[:0]`
Так явно очищается слайс, при этом не происходит нового выделения памяти.

applejohnny commented on internal/saver/saver.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// flushes a slice of Requests and makes a new empty slice with a certain capacity
+func (s *saver) flush(requests []models.Request) []models.Request {
```
Странный контракт функции, чтобы не было получено на вход - выход всегда будет одинаковый.
К тому же ф-ия делает сразу две вещи: сброс данных и создание нового слайса.
Имхо, `make([]models.Request, 0, s.capacity)`  стоит убрать и явно очищать слайс запросов.

applejohnny commented on internal/saver/saver.go:
```diff
@@ -0,0 +1,131 @@
+
+// flushes a slice of Requests and makes a new empty slice with a certain capacity
+func (s *saver) flush(requests []models.Request) []models.Request {
+	if len(requests) == 0 {
```
Проверка выглядит как лишняя, этот кейс должен быть обработан в `s.flusher.Flush(requests)`

vershininm committed:
task-5 several fixes per review:
- use bit literals instead of iota as state flag constant
- explicitly clear requests slice instead of making new one
- use defer to Done() wait group
